File: 0x08-python-more_classes/3-rectangle.py
#!/usr/bin/python3
""" Contains a rectangle class. """


class Rectangle:
    """ Creates a rectangle object. """
    def __init__(self, width=0, height=0):
        """ Initializes the rectangle object. """
        self.width = width
        self.height = height

    @property
    def width(self):
        """ Retrieves the rectangle's width. """
        return self.__width

    @property
    def height(self):
        """ Retrieves the rectangle's height. """
        return self.__height

    @width.setter
    def width(self, value):
        """ Sets the rectangle's width. """
        if not isinstance(value, int):
            raise TypeError('width must be an integer')
        if value < 0:
            raise ValueError('width must be >= 0')

        self.__width = value

    @height.setter
    def height(self, value):
        """ Sets the rectangle's height. """
        if not isinstance(value, int):
            raise TypeError('height must be an integer')
        if value < 0:
            raise ValueError('height must be >= 0')

        self.__height = value

    def area(self):
        """ Calculates and returns the area of the rectangle. """
        return self.__width * self.__height

    def perimeter(self):
        """ Calculates and returns the perimeter of the rectangle. """
        if self.__width == 0 or self.__height == 0:
            return 0
        return 2 * (self.__width + self.__height)

    def __str__(self):
        if self.__width == 0 or self.__height == 0:
            return ""
        shape = ""
        for y in range(self.__height):
            for x in range(self.__width):
                shape += "#"
            shape += "\n"

        return shape

File: 0x13-javascript_objects_scopes_closures/3-rectangle.js
#!/usr/bin/node
class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let square = '';
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        square += 'X';
      }
      if (i < this.height - 1) square += '\n';
    }
    console.log(square);
  }
}

module.exports = Rectangle;
